### PR TITLE
Ajaxify header

### PR DESCRIPTION
**Pull request by wjordan**
Rebased #1233 for merge into staging

### CODE REVIEW

**Comment by joshlory** [comment]
Trophies are missing from the header of http://localhost:3000/s/1/level/90

File: dashboard/app/controllers/reports_controller.rb [hunk 1]
```diff
@@ -29,6 +30,28 @@ def header_stats
     render file: 'shared/_user_stats', layout: false, locals: {user: current_user}
   end
 
+  def user_progress
+
+    # TODO: What kind of errors should we be throwing?
+    script_name = params[:script_name]
+    stage_id = params[:stage_id]
+    level_id = params[:level_id]
+    raise ActiveRecord::RecordNotFound unless script_name && level_id
```
Comment (joshlory) [inline dashboard/app/controllers/reports_controller.rb]: Why raise an exception here instead of returning 404?  (Two more below.)

File: dashboard/app/assets/javascripts/header.js.erb [hunk 2]
```diff
@@ -2,110 +2,575 @@
...
+        return this.resolve(args);
+
+      // Offline case
+      return {
+        url: this.resolveFilename(args, dataType)
+      };
+    }
+
+    UIStore.prototype.onAjaxFailure = function(xhr, status, message) {
+      // TODO
```
Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: TODO.  Do we plan on ever handling the AJAX failure?  If so, let's implement it.  If not, let's remove.

File: dashboard/app/assets/javascripts/header.js.erb [hunk 3]
```diff
@@ -2,110 +2,575 @@
...
+          stageHtml: html
+        });
+      });
+
+      // TODO: Handle props changing.
+      if (this.props.script_name)
+        this.headerStore.load( this.props );
+
+      // Create tooltips for the free-play icons
+      // TODO: Might be better to initialize this component elsewhere
```
Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: TODO

File: dashboard/app/assets/javascripts/header.js.erb (hunk 2 above)
Comment (SteveEisner) [inline dashboard/app/assets/javascripts/header.js.erb]: This is a base class.  We don't need to implement a fallback here, but this method is here to be overridden.

File: dashboard/app/assets/javascripts/header.js.erb (hunk 3 above)
Comment (SteveEisner) [inline dashboard/app/assets/javascripts/header.js.erb]: So far there's no better place, unfortunately.  As we create a better client shell, there will be!

**Comment by joshlory** [comment]
Nit: `var example = function () {...};` should have semicolons at the end.  `function example() {...}` is ok with no semicolon.  Minor style nit, doesn't block check-in.

File: dashboard/app/assets/javascripts/header.js.erb [hunk 4]
```diff
@@ -2,110 +2,575 @@
...
+// New private scope -- a lot of this is boilerplate for class structures, taken from coffeescript
+(function() {
+  var
+    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
+    __hasProp = {}.hasOwnProperty,
+    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
+    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };
+
+  // class UIComponent: this is a lightweight inheritable class modeled after React
+  UIComponent = (function() {
```
Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: `UIComponent` is implicitly defined (no `var`) so it will leak to global scope.  I think it's safe to add `var` to all of these to keep them private since they're only referenced inside this closure.  (A few more occurrences below.)

File: dashboard/app/assets/javascripts/header.js.erb [hunk 5]
```diff
@@ -2,110 +2,575 @@
...
+        // Allow things to settle - no need to call more than once per execution
+        this.isUpdating = setTimeout( this.onUpdate, 1 );
+      }
+
+      this.state = state || {};
+    }
+
+    // My state has been changed, and we've let all other processes yield, so we know
+    // there are no more changes coming.  Time to update the DOM.
+    UIComponent.prototype.onUpdate = function(ev) {
```
Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: Nit: `ev` event arg isn't used.

Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: Ah never mind, I see that you expect this to be overridden by the actual impl.  Makes sense to keep this for code completion.

File: dashboard/app/controllers/reports_controller.rb (hunk 1 above)
Comment (joshlory) [inline dashboard/app/controllers/reports_controller.rb]: Followed up offline, this is an unrelated bug in our routing.  ActiveRecord::RecordNotFound is the correct way to signal 404 here.

File: dashboard/app/assets/javascripts/header.js.erb (hunk 4 above)
Comment (SteveEisner) [inline dashboard/app/assets/javascripts/header.js.erb]: We can do that.  I actually do explicitly put them in window state further down, but it's correct that they should probably have var.  I grabbed the template from coffeescript's class transpile so I guess coffeescript makes all classes global...

File: dashboard/app/assets/javascripts/header.js.erb (hunk 2 above)
Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: Can the TODO be removed then?  It doesn't sound like there's any action needed.

File: dashboard/app/assets/javascripts/header.js.erb [hunk 6]
```diff
@@ -2,110 +2,575 @@
...
+
+    function UIRouter() { }
+
+    UIRouter.prototype.route = function() {
+      var re, parts,
+        url = window.location.pathname;
+
+      // s/:script_id/level/:level_id
+      // (only applies to the 20-hour course)
+      re = /\/s\/\d+\/level\/(\d+)$/i;
```
Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: Nit: I'd prefer to just inline both these regexes but no biggie

File: dashboard/app/assets/javascripts/header.js.erb [hunk 7]
```diff
@@ -2,110 +2,575 @@
...
+      return {
+        url: this.resolveUrl(args, dataType),
+        data: args
+      };
+    }
+
+    // What are the AJAX parameters for this callback?
+    // If you override this, the override should be offline-aware.
+    UIStore.prototype.resolveAjax = function(args, dataType) {
+      var ajax;
```
Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: `ajax` is unused

File: dashboard/app/assets/javascripts/header.js.erb (hunk 4 above)
Comment (joshlory) [inline dashboard/app/assets/javascripts/header.js.erb]: Wow that's wild, I wonder why.  What's the point of having a closure if the class definitions get leaked anyway?  :)

**Comment by joshlory** [comment]
Per our offline discussion, I would ignore most of these style nits since the eventual plan is to transition this code to CoffeeScript.

**Comment by SteveEisner** [comment]
I would love to, & hope to convince people of the value of CS :)  Maybe a topic for 1-2 weeks from now as I present some comparisons

**Comment by pcardune** [comment]
This can be closed, yes?

**Comment by wjordan** [comment]
I would assume this PR is now obsolete (since we've now integrated React into our Dashboard JS pipeline)- @joshlory @islemaster @Bjvanminnen could one of you confirm whether this PR can be closed?

**Comment by Bjvanminnen** [comment]
Yeah, this is outdated enough that I suspect it's not overly useful anymore.
